### PR TITLE
Revision models.detection.yolo 

### PR DESCRIPTION
**Pull request by heimish-kyma**
## What does this PR do?

Related to #839 

- [x] pl_bolts.models.detection.yolo.yolo_config.py
- [x] pl_bolts.models.detection.yolo.yolo_module.py
- [x] pl_bolts.models.detection.yolo.yolo_layers.py

Added some unit tests & yolo_giou config
- [x] tests.models.yolo.test_yolo_config.py
- [x] tests.models.yolo.test_yolo_layers.py
- [x] tests.data.yolo_giou.cfg

Feedback is welcomed with open arms.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does **only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? \[not needed for typos/docs\]
- [x] Did you verify new and **existing tests** pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review

- [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by heimish-kyma — COMMENTED** [review]
remove unused import & variable

File: tests/models/yolo/unit/test_yolo_config.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    assert conv.conv.stride == (config["stride"], config["stride"])
+
+    activation = config["activation"]
+    pad_size = (config["size"] - 1) // 2 if config["pad"] else 0
+
+    if config["pad"]:
+        assert conv.conv.padding == (pad_size, pad_size)
+
+    if config["batch_normalize"]:
+        assert len(conv) == 3
```
Comment (luca-medeiros) [inline tests/models/yolo/unit/test_yolo_config.py:38]: What is this part testing?

File: tests/models/yolo/unit/test_yolo_config.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    "config",
+    [
+        ({"batch_normalize": 1, "filters": 8, "size": 3, "stride": 1, "pad": 1, "activation": "leaky"}),
+        ({"batch_normalize": 0, "filters": 2, "size": 1, "stride": 1, "pad": 1, "activation": "mish"}),
+        ({"batch_normalize": 1, "filters": 6, "size": 3, "stride": 2, "pad": 1, "activation": "logistic"}),
+        ({"batch_normalize": 0, "filters": 4, "size": 3, "stride": 2, "pad": 0, "activation": "linear"}),
+    ],
+)
+def test_create_convolutional(config):
+    conv = _create_convolutional(config, [3])[0]
```
Comment (luca-medeiros) [inline tests/models/yolo/unit/test_yolo_config.py:25]: I believe changing to `conv, _ = _create_convolutional(config, [3])` would increase readability a bit.
Same for all the other `_create_*[0]` methods.

File: tests/models/yolo/unit/test_yolo_layers.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        (SELoss, [[0.0, 0.0, 120.0, 200.0]], [[189.0, 93.0, 242.0, 215.0]], 59479.0),
+    ],
+)
+def test_loss_functions(loss_func, bbox1, bbox2, expected):
+    loss_func = loss_func()
+    tensor1 = torch.tensor(bbox1, dtype=torch.float32)
+    tensor2 = torch.tensor(bbox2, dtype=torch.float32)
+
+    loss = loss_func(tensor1, tensor2)
+    assert loss.item() > 0.0
```
Comment (luca-medeiros) [inline tests/models/yolo/unit/test_yolo_layers.py:35]: For the given tests this might hold true, but GIoULoss(tensor1, tensor1) would be 0.0. Maybe being more explicit about how the loss cannot be negative would make more sense

**Review by luca-medeiros — COMMENTED** [review]
Very thoughtful tests! I would re-check all type hints once more and make sure tests comply with #844. Good stuff 👍

**Comment by luca-medeiros** [comment]
Maybe we can consider moving with https://github.com/Lightning-AI/lightning-bolts/pull/817 too.

File: tests/models/yolo/unit/test_yolo_config.py (hunk 1 above)
Comment (heimish-kyma) [reply tests/models/yolo/unit/test_yolo_config.py:38]: If the conv layer doesn't use batch normalize, the length of module is 2 which is consist of `conv and act` layers.

**Review by luca-medeiros — COMMENTED** [review]
You can check https://github.com/Lightning-AI/lightning-bolts/pull/848 to have an idea of how to use the 'catch_warnings' on all your tests.

File: tests/models/yolo/unit/test_yolo_layers.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+@pytest.mark.parametrize(
+    "loss_func, bbox1, bbox2, expected",
+    [
+        (GIoULoss, [[0.0, 0.0, 120.0, 200.0]], [[189.0, 93.0, 242.0, 215.0]], 1.4144532680511475),
+        (IoULoss, [[0.0, 0.0, 120.0, 200.0]], [[189.0, 93.0, 242.0, 215.0]], 1.0),
+        (SELoss, [[0.0, 0.0, 120.0, 200.0]], [[189.0, 93.0, 242.0, 215.0]], 59479.0),
+    ],
+)
+def test_loss_functions(loss_func, bbox1, bbox2, expected):
```
Comment (otaj) [inline tests/models/yolo/unit/test_yolo_layers.py:29]: Please, add `catch_warning` fixture.

File: tests/models/yolo/unit/test_yolo_layers.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+
+@pytest.mark.parametrize(
+    "xy, wh, expected",
+    [
+        ([0.0, 0.0], [1.0, 1.0], [-0.5, -0.5, 0.5, 0.5]),
+        ([5.0, 5.0], [2.0, 2.0], [4.0, 4.0, 6.0, 6.0]),
+    ],
+)
+def test_corner_coordinates(xy, wh, expected):
```
Comment (otaj) [inline tests/models/yolo/unit/test_yolo_layers.py:14]: Please, add `catch_warning` fixture.

File: tests/models/yolo/unit/test_yolo_config.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+@pytest.mark.parametrize(
+    "config",
+    [
+        ({"batch_normalize": 1, "filters": 8, "size": 3, "stride": 1, "pad": 1, "activation": "leaky"}),
+        ({"batch_normalize": 0, "filters": 2, "size": 1, "stride": 1, "pad": 1, "activation": "mish"}),
+        ({"batch_normalize": 1, "filters": 6, "size": 3, "stride": 2, "pad": 1, "activation": "logistic"}),
+        ({"batch_normalize": 0, "filters": 4, "size": 3, "stride": 2, "pad": 0, "activation": "linear"}),
+    ],
+)
+def test_create_convolutional(config):
```
Comment (otaj) [inline tests/models/yolo/unit/test_yolo_config.py:20]: Please, add `catch_warning` fixture.

File: tests/models/yolo/unit/test_yolo_config.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        ),
+        (
+            {
+                "size": 6,
+                "stride": 3,
+            }
+        ),
+    ],
+)
+def test_create_maxpool(config):
```
Comment (otaj) [inline tests/models/yolo/unit/test_yolo_config.py:60]: Please, add `catch_warning` fixture.

File: tests/models/yolo/unit/test_yolo_config.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+
+@pytest.mark.parametrize(
+    "config",
+    [
+        ({"from": 1, "activation": "linear"}),
+        ({"from": 3, "activation": "linear"}),
+    ],
+)
+def test_create_shortcut(config):
```
Comment (otaj) [inline tests/models/yolo/unit/test_yolo_config.py:76]: Please, add `catch_warning` fixture.

File: tests/models/yolo/unit/test_yolo_config.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+
+@pytest.mark.parametrize(
+    "config",
+    [
+        ({"stride": 2}),
+        ({"stride": 4}),
+    ],
+)
+def test_create_upsample(config):
```
Comment (otaj) [inline tests/models/yolo/unit/test_yolo_config.py:89]: Please, add `catch_warning` fixture.

File: tests/models/test_detection.py [hunk 10]
```diff
@@ -79,17 +79,25 @@ def test_fasterrcnn_pyt_module_bbone_train(tmpdir):
     trainer.fit(model, train_dl, valid_dl)
 
 
-def test_yolo(tmpdir):
-    config_path = Path(TEST_ROOT) / "data" / "yolo.cfg"
+@pytest.mark.parametrize("config", [("yolo"), ("yolo_giou")])
+def test_yolo(config):
```
Comment (otaj) [inline tests/models/test_detection.py:83]: Please, add `catch_warning` fixture.

File: tests/models/test_detection.py [hunk 11]
```diff
@@ -88,8 +88,15 @@ def test_yolo(tmpdir):
...
-def test_yolo_train(tmpdir):
-    config_path = Path(TEST_ROOT) / "data" / "yolo.cfg"
+@pytest.mark.parametrize(
+    "cfg_name",
+    [
+        ("yolo"),
+        ("yolo_giou"),
+    ],
+)
+def test_yolo_train(tmpdir, cfg_name):
```
[the remark this answers is not in the document]
Comment (otaj) [reply tests/models/test_detection.py:99]: Please, add `catch_warning` fixture.

**Review by otaj — COMMENTED** [review]
Hi, @heimish-kyma, sorry for taking so long. The PR in general looks good, I agree with all of the luca-medeiros's comments. (redundant argument in couple places, not using `catch_warnings` fixture.

However, if I can ask for a bit more careful review of `yolo_module.py` file? First of all, I'd like to avoid using mutable objects as default arguments to methods as it can very easily come back and bite you. Another nitpick is that in `infer()` you are setting the model to `eval()` mode, but not back to the original mode after the method.

Plus, to be fair, I am not very big fan of the design of `forward` method (one big opaque `ModuleList`, that is always doing same checks such as which module is supposed to have which inputs, every single time in forward instead of doing this once in `__init__` to have this constructed a bit more carefully. But, I absolutely understand you're not the original author of this implementation, so those complete redesigns might be a bit more difficult - and they're not the point of this review. But it's something I had to get off my chest :joy:

**Review by otaj — COMMENTED** [review]
Btw, `mypy is showing some serious errors in typing. We don't necessarily need to get typign to 100%, but we don't want to introduce wrong typing

File: pl_bolts/models/detection/yolo/yolo_config.py [hunk 12]
```diff
@@ -264,8 +255,10 @@ def _create_yolo(config, num_inputs):
         overlap_loss_func = yolo_layers.SELoss()
     elif overlap_loss_name == "giou":
         overlap_loss_func = yolo_layers.GIoULoss()
-    else:
+    elif overlap_loss_name == "iou":
         overlap_loss_func = yolo_layers.IoULoss()
+    else:
+        assert False, "Unknown overlap loss: " + overlap_loss_name
```
[the remark this answers is not in the document]
Comment (otaj) [reply pl_bolts/models/detection/yolo/yolo_config.py:261]: According to the code six lines above, the default should be `mse`

**Comment by otaj** [comment]
Hi, @heimish-kyma, do you plan to continue on this PR? :zap: Thank you! :nut_and_bolt:

**Comment by heimish-kyma** [comment]
Hi, @otaj sorry for being delayed first of all. 

I will take care things mentioned above comments this weekends, Cheers!

**Comment by otaj** [comment]
> Hi, @otaj sorry for being delayed first of all.
> 
> I will take care things mentioned above comments this weekends, Cheers!

@heimish-kyma, no need to apologize, I know things are taking a long time for me as well :smile: Thank you :zap:

**Review by otaj — COMMENTED** [review]
FIx most obvious failing tests

File: pl_bolts/models/detection/yolo/yolo_config.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (senarvi) [reply pl_bolts/models/detection/yolo/yolo_config.py:261]: This was based on the original YOLO implementation, which I think uses MSE loss. That's why "mse" was the default. Darknet configuration files allow other variants of the iou loss (diou, ciou) that were not available in Torchvision at the time of writing this code, so I chose to use IoULoss in case something else than "mse" or "giou" is specified in the configuration file. I created [another pull request](https://github.com/Lightning-AI/lightning-bolts/pull/817) a while ago that adds support for the ciou and diou losses, that are now available in Torchvision.

**Comment by senarvi** [comment]
> Maybe we can consider moving with #817 too.

@luca-medeiros I'm the author of that update. I'm sorry I didn't notice this pull request earlier. This PR seems to mostly contain new tests, which is very good, because the tests will always be useful. Otherwise, the YOLO code is old and has been improved quite a bit in my pull request. Better documented, complete type hints throughout the code, etc. I've also refactored it into multiple classes and added new features. Does it make sense to merge the tests from this PR and then move to https://github.com/Lightning-AI/lightning-bolts/pull/817 ? I mean, will the work be obsolete, if a lot of work is put into fixing things like the type hints in this PR?

File: pl_bolts/models/detection/yolo/yolo_layers.py [hunk 13]
```diff
@@ -500,5 +500,5 @@ def __init__(self, source_layer: int) -> None:
         super().__init__()
         self.source_layer = source_layer
 
-    def forward(self, x, outputs):
+    def forward(self, x, outputs: List[Union[Tensor, None]]) -> Tensor:
```
[the remark this answers is not in the document]
Comment (senarvi) [reply pl_bolts/models/detection/yolo/yolo_layers.py:506]: @luca-medeiros you're right. RouteLayer and ShortcutLayer do not use `x` (the output from the previous layer) and [calling them](https://github.com/Lightning-AI/lightning-bolts/blob/master/pl_bolts/models/detection/yolo/yolo_module.py#L153) is anyway handled as a special case, so the `x` can be dropped.

**Comment by senarvi** [comment]
> However, if I can ask for a bit more careful review of `yolo_module.py` file? First of all, I'd like to avoid using mutable objects as default arguments to methods as it can very easily come back and bite you. Another nitpick is that in `infer()` you are setting the model to `eval()` mode, but not back to the original mode after the method.
> 
> Plus, to be fair, I am not very big fan of the design of `forward` method (one big opaque `ModuleList`, that is always doing same checks such as which module is supposed to have which inputs, every single time in forward instead of doing this once in `__init__` to have this constructed a bit more carefully. But, I absolutely understand you're not the original author of this implementation, so those complete redesigns might be a bit more difficult - and they're not the point of this review. But it's something I had to get off my chest 😂

Hi @otaj,
good comments. I'm the original author, so I guess other than the `catch_warnings` fixture, these comments are related to my code. I fixed those small things in my [pull request](https://github.com/Lightning-AI/lightning-bolts/pull/817), i.e.

- removed unused module arguments,
- avoided mutable objects as default arguments, and
- infer() returns the model to the previous mode.

About the last comment: The design is ugly, because we read a Darknet configuration file and construct the network dynamically. My pull request adds support for using any typical PyTorch network that constructs the modules in `__init__`. That is, I have separated the Darknet network, which creates a `ModuleList` based on a Darknet configuration file, and the YOLO model, which can also accept traditional PyTorch networks.

If you have time to look at that pull request too, I would be interested to hear what you think.

**Comment by otaj** [comment]
@senarvi that is great to hear! If you could please merge your PR into this branch, that would be great! I would do it myself, but I have to say, I got a bit overwhelmed by the amount of changes in #817 and navigating them myself without knowing too much of the context felt like I could easily mess something up :joy: I believe that since you have more context about YOLO than I do, you'd be the perfect person to do so. Btw, nice tool you can use for it is `gh` (`github-cli`), that allows you to checkout a PR by number and registers branches and remotes for you.

Thank you!

**Comment by senarvi** [comment]
> @senarvi that is great to hear! If you could please merge your PR into this branch, that would be great!

Sure, I'll try to do that.

**Comment by senarvi** [comment]
@otaj I merged the changes, but as I guessed, I couldn't push them into this branch. I pushed them into my branch. So the merged changes appear now in [this pull request](https://github.com/Lightning-AI/lightning-bolts/pull/817). I don't know if that's what you meant. I'll now go through your comments there.

**Comment by senarvi** [comment]
@heimish-kyma should `self.log(..., batch_size=images.size(0))` be used in `test_step()` too?

File: pl_bolts/models/detection/yolo/yolo_config.py (hunk 12 above)
Comment (Borda) [inline pl_bolts/models/detection/yolo/yolo_config.py:261]:
```suggestion
        raise ValueError("Unknown overlap loss: " + overlap_loss_name)
```

**Review by Borda — APPROVED** [review]
